11-recommendation/helper_functions.py

```diff
@@ -0,0 +1,530 @@
+    GARDENING_SEEDS_PLANTS = "Gardening: Seeds and Plants"
+    GARDENING_SOIL_FERTILIZERS = "Gardening: Soil and Fertilizers"
+    GARDENING_GARDEN_DECOR = "Gardening: Garden Décor"
+    GARDENING_IRRIGATION_SUPPLIES = "Gardening: Irrigation Supplies"
+    CLEANING_CLEANING_SUPPLIES = "Cleaning: Cleaning Supplies"
+    CLEANING_BROOMS_MOPS = "Cleaning: Brooms and Mops"
+    CLEANING_VACUUM_CLEANERS = "Cleaning: Vacuum Cleaners"
+    CLEANING_PRESSURE_WASHERS = "Cleaning: Pressure Washers"
+    OUTDOOR_GRILLS_OUTDOOR_COOKING = "Outdoor: Grills and Outdoor Cooking"
+    OUTDOOR_LAWN_MOWERS = "Outdoor: Lawn Mowers"
+    OUTDOOR_PATIO_FURNITURE = "Outdoor: Patio Furniture"
+    OUTDOOR_FIRE_PITS = "Outdoor: Fire Pits"
+    SAFETY_PERSONAL_PROTECTIVE_EQUIPMENT = "Safety: Personal Protective Equipment"
+    SAFETY_FIRST_AID_SUPPLIES = "Safety: First Aid Supplies"
+    SAFETY_FIRE_EXTINGUISHERS = "Safety: Fire Extinguishers"
+    SAFETY_SAFETY_SIGNS = "Safety: Safety Signs"
+class Category(BaseModel):
+    item: CategoryEnum
+    description: str
+
+class CategoryRecommendations(BaseModel):
+    categories: List[Category]
+
+def expand_query_with_llm(query: str,  past_history: List[Dict[str, Any]]) -> str:
+    
+    """
+    Expands a user's query by leveraging OpenAI's language model to incorporate past history and categories.
+
+    :param query: The user's input message.
+    :param past_history: A list of dictionaries representing the user's past history.
+    :return: An expanded query string.
+    """
+    try:
+        # Prepare the prompt for the language model
+        # Format past purchases as per the expected template
+        past_purchases_entries = "".join([
+            f"<past_purchase>Item: {purchase['product_name']}</past_purchase>" for purchase in past_history
+        ])
+        system_prompt = f"""
+
+            You are an event planner with access to a product catalog.
+
+            Your task is to:
+
+            1. **Analyze the User's Input**: Based on the user's stated goals, identify up to 7 relevant unique categories from the provided list that align with their interests and needs. Ensure that they are relevant to the user's goals.
+
+            2. **Select Relevant Categories**:
+            - Only include categories that are directly relevant to the user's input.
+            - Bias towards categories that are complementary to the user's past purchases. However, if the user's goal is not related to their past purchases, do not include any categories related to their past purchases.
+
+            3. **Craft Engaging Descriptions**:
+            - For each selected category, write a fun and playful description to excite the user about their project.
+            - ONLY if applicable, mention how their past purchases contribute to their progress, highlighting that they're already on their way.
+
+            
+            **Best Practices**:
+            4. **Include Essential and Complementary Categories**:
```

Comment:
Not sure you need to number them anymore for best practices. You should just have it in a list for 56? Which parts are best practices?